### PR TITLE
Handle null user names

### PR DESCRIPTION
**Pull request by WiNloSt**
## Note
After https://github.com/metabase/metabase/pull/23154 is merged. `common_name` will fallback to user email.

related to #22754

### CODE REVIEW

File: frontend/src/metabase/admin/people/containers/UserPasswordResetModal.jsx [hunk 1]
```diff
@@ -44,10 +46,10 @@ class UserPasswordResetModal extends React.Component {
             ml="auto"
             onClick={async () => {
               if (emailConfigured) {
-                await user.passwordResetEmail();
+                await user.resetPasswordEmail();
```
Comment (WiNloSt) [inline frontend/src/metabase/admin/people/containers/UserPasswordResetModal.jsx:47]: Make action names start with verb.

File: frontend/src/metabase/admin/people/containers/UserSuccessModal.jsx [hunk 2]
```diff
@@ -24,7 +24,8 @@ class UserSuccessModal extends React.Component {
     const { onClose, user, temporaryPassword } = this.props;
     return (
       <ModalContent
-        title={t`${user.getName()} has been added`}
+        // https://user-images.githubusercontent.com/1937582/172612281-9cfd67a5-b228-4176-a64b-5ee3e2369179.png
+        title={t`${user.common_name} has been added`}
```
Comment (WiNloSt) [inline frontend/src/metabase/admin/people/containers/UserSuccessModal.jsx:27]: User's [`getName`](https://github.com/metabase/metabase/pull/23111/files#diff-4f46204c4fd684bc8b51d8c6f9a2d791ce065787b6de8dce6808772d1d02f5c8L33) is used sparingly, so it would be better that we drop this usage as most user props are just plain objects anyway.

File: frontend/src/metabase/admin/people/containers/UserSuccessModal.jsx [hunk 3]
```diff
@@ -41,7 +42,6 @@ class UserSuccessModal extends React.Component {
 export default _.compose(
   User.load({
     id: (state, props) => props.params.userId,
-    wrapped: true,
```
Comment (WiNloSt) [inline frontend/src/metabase/admin/people/containers/UserSuccessModal.jsx:44]: Since there's no need for user entity methods (`getName`), we can simply pass an object instead.

File: frontend/src/metabase/collections/components/BaseTableItem.jsx [hunk 4]
```diff
@@ -54,7 +54,6 @@ export function BaseTableItem({
   onMove,
   onDrop,
   onToggleSelected,
-  ...props
```
Comment (WiNloSt) [inline frontend/src/metabase/collections/components/BaseTableItem.jsx:57]: Not sure why this unused var isn't caught by our linter since this is a javascript file

File: frontend/src/metabase/collections/components/BaseTableItem.jsx [hunk 5]
```diff
@@ -49,12 +48,10 @@ export function BaseTableItem({
   isSelected,
   isPinned,
   linkProps = {},
-  hasBottomBorder = true,
```
Comment (WiNloSt) [inline frontend/src/metabase/collections/components/BaseTableItem.jsx:52]: Not sure why this unused var isn't caught by our linter since this is a javascript file

File: frontend/src/metabase/components/Header.jsx [hunk 6]
```diff
@@ -143,14 +143,9 @@ class Header extends Component {
       );
     }
 
-    let attribution;
-    if (this.props.item && this.props.item.creator) {
-      attribution = (
-        <div className="Header-attribution">
```
Comment (WiNloSt) [inline frontend/src/metabase/components/Header.jsx:149]: The class name `Header-attribution` is added in https://github.com/metabase/metabase/pull/2136. Alongside, CSS that hid it.
https://github.com/metabase/metabase/pull/2136/files#diff-70dba2e3e8a73fb80938e1c5e8414d87d5bac4bff80b00708420999b726b9f95R11-R15

So, if we don't render it (`display: none;`) we might as well remove it.

File: frontend/src/metabase/css/components/header.css [hunk 7]
```diff
@@ -7,12 +7,6 @@
   color: var(--color-text-dark);
 }
 
-.Header-attribution {
-  display: none; /* disabled */
-  color: var(--color-text-medium);
-  margin-bottom: 0.5em;
-}
```
Comment (WiNloSt) [inline frontend/src/metabase/css/components/header.css:14]: Based on [comment placeholder for frontend/src/metabase/components/Header.jsx], since its usage is removed, this should be removed as well.

File: frontend/src/metabase/components/LastEditInfoLabel.js [hunk 8]
```diff
@@ -32,19 +32,21 @@ LastEditInfoLabel.propTypes = {
...
-function formatEditorName(firstName, lastName) {
-  const lastNameFirstLetter = lastName.charAt(0);
-  return `${firstName} ${lastNameFirstLetter}.`;
+function formatEditorName(lastEditInfo) {
+  //  https://user-images.githubusercontent.com/1937582/172189537-0eb7f620-559d-4537-8467-5e1ea5bab78a.png
+  const name = [lastEditInfo.first_name, lastEditInfo.last_name]
+    .join(" ")
+    .trim();
+
+  return name || lastEditInfo.email;
```
Comment (WiNloSt) [inline frontend/src/metabase/components/LastEditInfoLabel.js:40]: Following [the discussion on Slack](https://metaboat.slack.com/archives/C03E8MQJZBM/p1654766630306639?thread_ts=1654759255.159039&cid=C03E8MQJZBM), this special one-off place should also be replaced with `common_name` but since `common_name` doesn't exist on `last-edit-info`, I need to manually fall back to email.

#### Edited by you


#### Edited by others
image

File: frontend/src/metabase/components/UserAvatar/UserAvatar.jsx [hunk 9]
```diff
@@ -29,7 +29,20 @@ function initial(name) {
 }
 
 function userInitials(user) {
-  return user ? initial(user.first_name) + initial(user.last_name) : null;
+  // https://user-images.githubusercontent.com/1937582/171615363-40bb26ca-6d81-4d0f-b144-2d679d2ede77.png
+  if (user) {
+    return nameInitials(user) || emailInitials(user);
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserAvatar/UserAvatar.jsx:33]: Following [the discussion on Slack](https://metaboat.slack.com/archives/C03E8MQJZBM/p1654766630306639?thread_ts=1654759255.159039&cid=C03E8MQJZBM), uses the first 2 letters from an email as an initial name.

File: frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+SingleName.args = {
+  user: {
+    first_name: "John",
+    last_name: null,
+    email: "john.w@hightable.org",
+    common_name: "John",
+  },
+};
+
+export const OnlyEmail = Template.bind({});
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx:34]: image

File: frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+OnlyEmail.args = {
+  user: {
+    first_name: null,
+    last_name: null,
+    email: "john.w@hightable.org",
+    common_name: "john.w@hightable.org",
+  },
+};
+
+export const shortEmail = Template.bind({});
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx:44]: image

File: frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+Default.args = {
+  user: {
+    first_name: "John",
+    last_name: "Wick",
+    email: "john.w@hightable.org",
+    common_name: "John Wick",
+  },
+};
+
+export const SingleName = Template.bind({});
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx:24]: image

File: frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+export default {
+  title: "Components/UserAvatar",
+  component: UserAvatar,
+};
+
+const Template: ComponentStory<typeof UserAvatar> = args => (
+  <UserAvatar {...args} />
+);
+
+export const Default = Template.bind({});
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx:14]: image

File: frontend/src/metabase/components/UserPicker/UserPicker.jsx [hunk 14]
```diff
@@ -24,15 +24,19 @@ const UserPicker = ({ value, validateValue, users, canAddItems, onChange }) => {
     : null;
 
   const options = useMemo(() => {
-    return users.map(user => ({ label: user.common_name, value: user }));
+    // `label` here isn't really used because we specify `filterOption`.
+    // Normally, `options` will be filtered by its `label` if we don't provide
+    // `filterOption` to <TokenField />.
+    return users.map(user => ({ value: user }));
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserPicker/UserPicker.jsx:27]: `label` here isn't really used because we specify `filterOption`. Normally, `options` will be filtered by its `label` if we don't provide `filterOption` to <TokenField />.

https://github.com/metabase/metabase/blob/d0ad25058d96436ee1995e43de4fbdc00c830ca5/frontend/src/metabase/components/TokenField/TokenField.tsx#L206-L209

File: frontend/src/metabase/components/UserPicker/UserPicker.jsx [hunk 15]
```diff
@@ -24,15 +24,19 @@ const UserPicker = ({ value, validateValue, users, canAddItems, onChange }) => {
...
   }, [users]);
 
   const idKey = useCallback(value => {
     return value.id || value.email;
   }, []);
 
   const valueRenderer = useCallback(value => {
-    return value.common_name || value.email;
+    // https://user-images.githubusercontent.com/1937582/172158045-802e7600-531d-4dd8-86f9-b4bd4dc936aa.png
+    return value.common_name;
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserPicker/UserPicker.jsx:35]: `common_name` already fall back to `email`

File: frontend/src/metabase/components/UserPicker/UserPicker.jsx [hunk 16]
```diff
@@ -78,8 +82,8 @@ const UserPicker = ({ value, validateValue, users, canAddItems, onChange }) => {
   );
 };
 
-const includesIgnoreCase = (s1, s2) => {
-  return s1.toLowerCase().includes(s2.toLowerCase());
+const includesIgnoreCase = (sourceText, searchText) => {
+  return sourceText.toLowerCase().includes(searchText.toLowerCase());
```
Comment (WiNloSt) [inline frontend/src/metabase/components/UserPicker/UserPicker.jsx:82]: Make the function similar to aother place and improve readability.

https://github.com/metabase/metabase/blob/ef9a0acc08640144a76c68769415d81e532d5d5e/frontend/src/metabase/admin/databases/components/widgets/EngineWidget/EngineWidget.tsx#L297-L302

File: frontend/src/metabase/pulse/components/RecipientPicker.jsx [hunk 17]
```diff
@@ -82,15 +86,9 @@ export default class RecipientPicker extends Component {
...
-              ~option.value.common_name
-                .toLowerCase()
-                .indexOf(filterString.toLowerCase()) ||
-              ~option.value.email
-                .toLowerCase()
-                .indexOf(filterString.toLowerCase())
-            }
+            // Just a note, but I think logic that filter options by either common_name or email might already work.
+            // But making this close to what we have in UserPicker would be better because that one is more readable.
+            filterOption={filterOption}
```
Comment (WiNloSt) [inline frontend/src/metabase/pulse/components/RecipientPicker.jsx:81]: Refactor this part to make it similar to another place and improve redability.

https://github.com/metabase/metabase/blob/6da7a41500daf73943b29797f486f9f91658ffa3/frontend/src/metabase/components/UserPicker/UserPicker.jsx#L51-L56

File: frontend/src/metabase/entities/users.js [hunk 18]
```diff
@@ -29,9 +29,7 @@ const Users = createEntity({
 
   path: "/api/user",
 
-  objectSelectors: {
-    getName: user => user.common_name || `${user.first_name} ${user.last_name}`,
-  },
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/users.js:34]: `getName` usage is sparse, so it's better be gone in favor of `common_name`

File: frontend/src/metabase/entities/users.js [hunk 19]
```diff
@@ -75,15 +73,15 @@ const Users = createEntity({
       await UserApi.send_invite({ id });
       return { type: RESEND_INVITE };
     },
-    passwordResetEmail: async ({ email }) => {
+    resetPasswordEmail: async ({ email }) => {
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/users.js:74]: Make action names start with verbs

File: frontend/src/metabase/entities/users/forms.js [hunk 20]
```diff
@@ -13,12 +13,14 @@ const getNameFields = () => [
     title: t`First name`,
     placeholder: "Johnny",
     autoFocus: true,
+    // XXX: Make the name optional after BE has relaxed name field constraint.
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/users/forms.js]: I left this in place, so I could be easily followed up in another PR, let me know if this shouldn't be here.

File: frontend/src/metabase/pulse/components/RecipientPicker.jsx [hunk 21]
```diff
@@ -73,7 +76,8 @@ export default class RecipientPicker extends Component {
             }
             autoFocus={autoFocus && recipients.length === 0}
             multi
-            valueRenderer={value => value.common_name || value.email}
+            // https://user-images.githubusercontent.com/1937582/172163846-86636488-9cb7-4b8f-9609-594b42384f4a.png
+            valueRenderer={value => value.common_name}
```
Comment (WiNloSt) [inline frontend/src/metabase/pulse/components/RecipientPicker.jsx:72]: `common_name` already falls back to `email`

File: frontend/src/metabase/components/UserAvatar/UserAvatar.jsx [hunk 22]
```diff
@@ -29,7 +29,20 @@ function initial(name) {
...
+
+  return null;
+}
+
+function nameInitials(user) {
+  return initial(user.first_name) + initial(user.last_name);
+}
+
+function emailInitials(user) {
+  const emailUsername = user.email.split("@")[0];
```
Comment (ranquild) [inline frontend/src/metabase/components/UserAvatar/UserAvatar.jsx:44]: Could `email` be null or an empty string?

**Review by ranquild — APPROVED** [review]
Looks good, please fix small issues mentioned in comments

File: frontend/src/metabase/home/homepage/components/HomeGreeting/HomeGreeting.tsx [hunk 23]
```diff
@@ -36,12 +36,13 @@ const HomeGreeting = ({
...
 
 const getMessage = (name: string): string => {
+  const namePart = name ? `, ${name}` : "";
   const options = [
-    t`Hey there, ${name}`,
-    t`How's it going, ${name}?`,
-    t`Howdy, ${name}`,
-    t`Greetings, ${name}`,
-    t`Good to see you, ${name}`,
+    t`Hey there` + namePart,
```
[the remark this answers is not in the document]
Comment (mazameli) [reply frontend/src/metabase/home/homepage/components/HomeGreeting/HomeGreeting.tsx:41]: It looks to me like both in the current change as well as in Alexander's suggestion there's a missing space between "there" and the namePart.

[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/home/homepage/components/HomeGreeting/HomeGreeting.tsx:41]: @mazameli That's because `namePart` could contain an optional comma so a single translation could result in 2 separate greetings
1. "Hey there": When names are null
2. "Hey there, Maz"

File: frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx (hunk 11 above)
Comment (WiNloSt) [reply frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx:44]: Dang. I missed that. Thank you!

File: frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+Default.args = {
+  user: {
+    first_name: "John",
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/components/UserAvatar/UserAvartar.stories.tsx:17]: Alright! I'll evacuate Mr. Wick right away!

File: frontend/src/metabase/components/UserAvatar/UserAvatar.jsx (hunk 22 above)
Comment (cdeweyx) [reply frontend/src/metabase/components/UserAvatar/UserAvatar.jsx:44]: Yeah it's required, afaik it's how we id customers

File: frontend/src/metabase/home/homepage/components/HomeGreeting/HomeGreeting.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (mazameli) [reply frontend/src/metabase/home/homepage/components/HomeGreeting/HomeGreeting.tsx]: Gotcha gotcha 👌🏼
